src/raftstore/store/read_queue.rs

```diff
@@ -0,0 +1,260 @@
+            }
+        }
+        RAFT_READ_INDEX_PENDING_COUNT.sub(removed as i64);
+        self.contexts.clear();
+        self.ready_cnt = 0;
+        self.handled_cnt = 0;
+    }
+
+    pub fn clear_uncommitted(&mut self, term: u64) {
+        let mut removed = 0;
+        for mut read in self.reads.drain(self.ready_cnt..) {
+            self.contexts.remove(&read.id);
+            removed += read.cmds.len();
+            for (_, cb) in read.cmds.drain(..) {
+                apply::notify_stale_req(term, cb);
+            }
+        }
+        RAFT_READ_INDEX_PENDING_COUNT.sub(removed as i64);
+    }
+
+    pub fn push_back(&mut self, read: ReadIndexRequest, is_leader: bool) {
+        if !is_leader {
+            let offset = self.handled_cnt + self.reads.len();
+            self.contexts.insert(read.id, offset);
+        }
+        self.reads.push_back(read);
+    }
+
+    pub fn back_mut(&mut self) -> Option<&mut ReadIndexRequest> {
+        self.reads.back_mut()
+    }
+
+    pub fn advance_leader_reads<T>(&mut self, states: T) -> Option<Timespec>
+    where
+        T: IntoIterator<Item = (Uuid, u64)>,
+    {
+        let mut ts = None;
+        for (uuid, index) in states {
+            assert_eq!(uuid, self.reads[self.ready_cnt].id);
+            self.reads[self.ready_cnt].read_index = Some(index);
+            ts = Some(self.reads[self.ready_cnt].renew_lease_time);
+            self.ready_cnt += 1;
+        }
+        ts
+    }
+
+    /// update the read index of the requests that before the specified id.
+    pub fn advance_replica_reads<T>(&mut self, states: T)
+    where
+        T: IntoIterator<Item = (Uuid, u64)>,
+    {
+        let (mut min_changed_offset, mut max_changed_offset) = (usize::MAX, 0);
+        for (uuid, index) in states {
+            if let Some(offset) = self.contexts.remove(&uuid) {
+                let offset = offset.checked_sub(self.handled_cnt).unwrap();
+                self.ready_cnt = cmp::max(self.ready_cnt, offset + 1);
+
+                if let Some(occur_index) = self.reads[offset].read_index {
+                    if occur_index < index {
+                        continue;
+                    }
+                }
+                self.reads[offset].read_index = Some(index);
+                min_changed_offset = cmp::min(min_changed_offset, offset);
+                max_changed_offset = cmp::max(max_changed_offset, offset);
+                continue;
+            }
+            debug!(
+                "cannot find corresponding read from pending reads";
+                "uuid" => ?uuid, "read-index" => index,
+            );
+        }
+
+        if self.ready_cnt > 1 && max_changed_offset > 0 {
+            self.fold(min_changed_offset, max_changed_offset);
+        }
+    }
+
+    pub fn fold(&mut self, mut min_changed_offset: usize, max_changed_offset: usize) {
+        let mut r_offset = max_changed_offset;
+        while r_offset >= min_changed_offset && r_offset > 0 {
```

Comment:
It feels weird to update both `r_offset` and `min_changed_offset` at the same time.